Feat: Sequelize 연결 설정과 엔티티 모델 구현, babel 설정 추가

changheedev opened the pull request:
## 작업내용

- Babel 설정 추가

  - 백엔드 코드에서도 프론트와 동일하게 import/export 패턴을 사용하기 위해 바벨 설정을 추가했습니다.

- DB 모델링
  - Sequelize 연결 설정

  - User, Issue, Milestone, Comment, Label 모델 생성 작업

  - 엔티티 모델 관계 설정

    - 일단은 모델이 어떻게 쓰일지 모르기 때문에 단방향 참조만 설정해두었습니다. 이후 필요에 따라 양방향으로 변경해서 사용하면 됩니다.

**생성된 db 결과**

![스크린샷 2020-10-28 오후 11 56 10](https://user-images.githubusercontent.com/17294694/97453553-340df800-1979-11eb-9893-82dfab042b84.png)

## 리뷰 요청

- Sequelize를 사용하면 모듈화를 어떻게 해야 할지 항상 고민인데 조언 부탁드립니다...

**모델을 사용하는 방식**

```javascript
import User from '../models/user';

//사용
```
```javascript
import {sequelize} from '../models';

const User = sequelize.models.User;
//사용        
```

**초기화 코드**

```javascript
const sequelize = new Sequelize(sequelizeConfigs);
const models = [User, Issue, Label, Milestone, Comment];
models.forEach((model) => model.init(sequelize));

Issue.belongsTo(User);
Issue.belongsTo(Milestone);
Comment.belongsTo(User);
Comment.belongsTo(Issue);
User.belongsToMany(Issue, { through: 'issue_assignees', timestamps: false });
Issue.belongsToMany(User, { through: 'issue_assignees', timestamps: false });
Issue.belongsToMany(Label, { through: 'issue_labels', timestamps: false });
Label.belongsToMany(Issue, { through: 'issue_labels', timestamps: false });
```

changheedev committed:
Feat: Sequelize 연결 설정

changheedev committed:
Feat: Sequelize 연결 설정

changheedev committed:
Chore: Babel 설정 추가

백엔드 코드에서도 프론트와 동일하게 import/export 패턴을 사용하기 위해 바벨 설정을 추가했습니다.

changheedev committed:
Chore: Babel 설정 추가

백엔드 코드에서도 프론트와 동일하게 import/export 패턴을 사용하기 위해 바벨 설정을 추가했습니다.

changheedev committed:
Feat: User Entity 구현

changheedev committed:
Feat: User Entity 구현

changheedev committed:
Feat: Issue Entity 구현

changheedev committed:
Feat: Issue Entity 구현

changheedev committed:
Feat: Label Entity 구현

changheedev committed:
Feat: Label Entity 구현

changheedev committed:
Feat: Milestone Entity 구현

changheedev committed:
Feat: Milestone Entity 구현

changheedev committed:
refactor: sequelize model 초기화 코드 수정

init을 직접 호출해주는 대신 model을 배열에 담고 forEach문을 통해 init이 호출되도록 수정

changheedev committed:
refactor: sequelize model 초기화 코드 수정

init을 직접 호출해주는 대신 model을 배열에 담고 forEach문을 통해 init이 호출되도록 수정

changheedev committed:
Chore: 루트 package.json에 서버 실행 스크립트 추가

개발 과정에서 서버 실행 명령어를 편하게 실행할 수 있도록 루트 디렉토리의 의 package.json에 스크립트를 추가했습니다.

changheedev committed:
Chore: 루트 package.json에 서버 실행 스크립트 추가

개발 과정에서 서버 실행 명령어를 편하게 실행할 수 있도록 루트 디렉토리의 의 package.json에 스크립트를 추가했습니다.

changheedev committed:
Chore: Webpack babel-loader 설정 추가

백엔드 코드에 babel을 적용했기 때문에 webpack에도 babel-loader 설정을 추가했습니다.

changheedev committed:
Chore: Webpack babel-loader 설정 추가

백엔드 코드에 babel을 적용했기 때문에 webpack에도 babel-loader 설정을 추가했습니다.

changheedev committed:
Feat: Comment Entity 구현

changheedev committed:
Feat: Comment Entity 구현

changheedev committed:
Feat: 엔티티 모델 관계설정

엔티티 모델의 관계를 설정했는데, 일단은 모델이 어떻게 쓰일지 모르기 때문에 단방향 참조만 설정해두었습니다. 이후 필요에 따라 양방향으로 변경해서 사용하면 됩니다.

changheedev committed:
Feat: 엔티티 모델 관계설정

엔티티 모델의 관계를 설정했는데, 일단은 모델이 어떻게 쓰일지 모르기 때문에 단방향 참조만 설정해두었습니다. 이후 필요에 따라 양방향으로 변경해서 사용하면 됩니다.

rolled-potatoes commented on server/src/models/label.js:
```diff
@@ -0,0 +1,33 @@
+  tableName: 'labels',
+};
+
+class Label extends Model {
```
이부분에서 관계를 지정 할 수 있는 것으로 알고 있는데 따로 모아둔 이유가 있을 까요?
```js
static associate(models){
  this.hasMany(models.other);
}
```

changheedev replied:
오홍 한번 적용해볼께요

yejineee commented on server/src/models/user.js:
```diff
@@ -0,0 +1,35 @@
+    allowNull: false,
+    primaryKey: true,
+    autoIncrement: true,
+  },
```
따로 id를 지정해주지 않아도, INTEGER로 자동으로 만들어주는 걸로 알고 있습니다..! 참고용...ㅎ

yejineee commented on server/src/models/user.js:
```diff
@@ -0,0 +1,35 @@
+  profile_link: {
+    type: DataTypes.STRING(255),
+    allowNull: false,
+  },
```
type에서 new를 붙인 것과 안 붙인 것의 차이가 있을까요?

changheedev replied:
확인해보고 차이가 없으면 하나로 통일시키겠습니다 👍  놓치고 있었네요

LeeSuKyeong replied:
new를 붙이는 이유가 무엇인가요?

changheedev replied:
제가 예전에 구현을 뒤죽박죽으로 해놓고 계속 참고해서 쓰고 있었네요. 
메뉴얼 보니까 new 없이 쓰고 있는데 new 없는 걸로 통일 시키겠습니다.

LeeSuKyeong committed:
Feat: CustomButton 추가, ColorStyle 추가 (#33)

* Feat : colorStyle 추가

button 테마색을 미리 지정하기 위해서 colorStyle.js 추가.

* Feat: CustomButton 추가

버튼UI 재사용을 위한 customButton을 생성.

* Feat: colorStyle에 borderColor 추가

테마에 따라 테두리 색상이 달라서 변경할 필요성을 느껴 prototype을 추가했습니다.

yejineee committed:
이슈 리스트 페이지에서 서버에 api 요청 및 useReducer로 상태 관리 (#34)

* Feat: 이슈 리스트 컴포넌트 생성 (#30)

* Refactor: App.js에서 가장 첫 페이지를 이슈 리스트 페이지로 설정

로그인 후, 가장 먼저 보이는 페이지를 이슈 리스트 페이지이므로
해당 페이지 컴포넌트를 가져와서 렌더링하는 것으로 수정하였다.

* Style: 이슈 리스트에서 쓰이는 컬러 추가

이슈 리스트에서 쓰이는 파란색 3개를 추가하였다.

* Feat: 리스트 헤더 컴포넌트 추가

리스트 헤더는 라벨 페이지, 이슈 리스트 페이지, 마일스톤 리스트 페이지 등
여러 페이지의  헤더에 공통적으로 쓰인다. 따라서 스타일을 지정한 리스트 컴포넌트를
추가하였다.

* Feat: 이슈 리스트 컨테이너 컴포넌트 추가

첫 페이지의 이슈 리스트 헤더와 목록들을 보여주는
컴포넌트를 추가하였다.
페이지부터, setState함수를 가져와서 이슈  헤더 클릭 이벤트와
이슈 바디 클릭 이벤트 핸들러를 따로 정의하였다.
이렇게 함으로써, 어떤 버튼이 클릭되었는지에 대한 분기 처리를 하지
않아도 되도록 하였다.

* Feat: 이슈 리스트 페이지 컴포넌트 추가

이슈 리스트 페이지 컴포넌트는 이슈 필터 컴포넌트와, 이슈 리스트 컴포넌트를
자식으로 갖고 있다.
두 컴포넌트에서 공통으로 state를 사용하기 때문에, 이 페이지 컴포넌트에서
state를 초기화하고, state와 setState 함수를 자식에게 넘겨주도록 한다.

Co-authored-by: yejineee <21700431@handong.edu>

* Feat : issue 더미 api 생성 (#31)

임시로 보여줄 데이터를 json으로 읽어와서 반환 함.

* Docs : issue 더미 api 명세서 작성

* Chore: axios baseURL 설정 및 response interceptor 설정

- 서버의 주소는 매 번 반복되므로, 오류를 줄이고, 편리함을 높이기 위해
axios의 instance를 생성하여 기본 URL을 설정하였다.

- axios의 response에서 매 번 data를 받아와야 하므로, interceptor하여
data를 바로 넘겨줄 수 있도록 설정하였다.

* Chore: 모든 이슈를 서버로부터 받아오는 axios 설정

서버로부터 api요청하는 코드를 apis 폴더에 저장한다.
그 중 api/issue.js는 이슈 모델과 관련하여 api요청하는 코드를 모아둘 것이다.
이슈의 모든 모델들을 가져오는 axios를 추가하였다.

* Chore: issue의 useReducer에서 사용되는 action type 상수 저장

useReducer에서 사용되는 action의 type을 상수로 지정하여,
코드 오류를 줄이고, 생산성을 높이도록 하였다.

* Feat: issue의 useReducer에서 쓰이는 reducer 정의

issue 페이지에서 쓰일 reducer를 정의하였다.
- FETCH_SUCCESS:
    - fetch 성공시 issues state를 새롭게 정의한다.
- FETCH_ERROR:
    - Fetch에 실패시, error state를 정의한다.
- CHECK_ALL_ISSUE_HANDLER:
    - 이슈 헤더에 있는 체크박스 클릭 시, 전체 issue 상태를 바꿔준다.
- CHECK_ISSUE_HANDLER:
    - 각 이슈 옆에 있는 체크박스 클릭 시, 해당 issue의 상태를 바꿔주고, 필요시 전체 issue의 상태를 바꿔준다.

* Refactor: 이슈 리스트 컴포넌트의 상태관리 방법을 useReducer로 바꿈

useState로 상태를 관리할 경우, 여러 컴포넌트에서 상태를 변경하게 되어
한 눈에 알아보기 힘든 점이 있었다.
또한, 이슈 리스트 페이지에서 api요청을 한 후 상태 변경을 useReducer로 하기 때문에, 기존에 useState로 관리하던 것을 useReducer로 바꿔줘야 했다.
따라서 IssueListContainer안에 있는 컴포넌트에서 클릭 이벤트 발생 시
상태 변경을 reducer가 하도록 수정하였다.

* Feat: IssueListPage에서 api요청하여 데이터 받는 것 구현

IssueListPage에서 api요청으로 issue들을 받아서, 자식 컴포넌트에게
전달하여 렌더링하도록 해야 한다.
따라서 api를 요청한 후, issue들을 받고, 상태 변경을 useReducer를 사용하도록 하였다.

* Chore: axios 설치

서버에 api요청 시 사용하는 라이브러리로 axios를 선택하여, 설치하였다.

* Chore: 프론트 eslint 규칙 추가

* Style: map에서 중괄호 삭제, 리턴 받는 변수명 변경

좀 더 명확한 네이밍을 위해, 서버로 부터 받아온 데이터를 issues라고
명명하였다.
map에서 하나의 statement만 있을 때, 중괄호를 삭제할 수 있으므로 삭제하였다.

Co-authored-by: yejineee <21700431@handong.edu>
Co-authored-by: ParkSeungHwan <44409642+rolled-potatoes@users.noreply.github.com>

changheedev committed:
Refactor: 엔티티 관계 설정 코드 개선

승환님 리뷰를 참고하여 엔티티의 관계를 설정하는 코드를 각 엔티티 모델 내부에 위치하도록 변경했습니다.
초기화 부분 코드가 훨씬 깔끔해졌네요.

changheedev committed:
Refactor: 엔티티 초기화 코드 개선

기존에는 관계설정을 하기위해 모든 모델을 index.js에서 import 했었는데
관계설정 코드를 수정한 이후에는 모델을 직접 참조하는 코드가 없어서 fs를 이용해
불러오는 방식으로 개선했습니다.

changheedev committed:
Merge branch 'feature/models' of https://github.com/changheedev/IssueTracker-25 into feature/models

# Conflicts:
#	server/src/app.js
#	server/src/models/comment.js
#	server/src/models/index.js
#	server/src/models/issue.js
#	server/src/models/label.js
#	server/src/models/milestone.js
#	server/src/models/user.js
